File: Data Structures & Algorithms/combination-target-sum/submission-9.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> subset;
        dfs(0, subset, 0, target, nums);
        return res;
    }
    void dfs(int idx, vector<int> subset, int count, int target, vector<int>& nums) {
        if (count == target) {
            res.push_back(subset);
            return;
        } else if (idx >= nums.size() || count > target) {
            return;
        }

        subset.push_back(nums[idx]);
        dfs(idx, subset, count + nums[idx], target, nums);
        subset.pop_back();
        dfs(idx + 1, subset, count, target, nums);
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-9.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        self.res = []
        self.cache = {}
        candidates.sort()

        def dfs(i, subset, val):
            if val == target:
                self.res.append(subset.copy())
                return

            if i >= len(candidates) or val > target:
                return
            
            subset.append(candidates[i])
            dfs(i + 1, subset, val + candidates[i])
            subset.pop()

            while i + 1 < len(candidates) and candidates[i] == candidates[i + 1]:
                i += 1

            dfs(i + 1, subset, val)
        
        dfs(0, [], 0)

        return self.res